cmds/tail/tail.go

```diff
@@ -0,0 +1,259 @@
+			idx = bytes.LastIndexByte(slice, '\n')
+			if idx == -1 {
+				// there are less than `numLines` lines
+				break
+			}
+			foundLines++
+			slice = slice[:idx-1]
+		}
+		if idx == -1 {
+			// if there are less than `numLines` lines, use all what we have read
+			data = readData
+		} else {
+			data = readData[idx+1:] // +1 to skip the newline belonging to the previous line
+		}
+	}
+	// write the requested lines to the writer
+	if _, err = writer.Write(data); err != nil {
+		return err
+	}
+	// reposition the stream at the end, so the caller can keep reading the file
+	// (e.g. when using follow-mode)
+	if _, err = inFile.Seek(lastPos, os.SEEK_SET); err != nil {
+		return err
+	}
+	return nil
+}
+
+// readLastLinesFromBeginning reads the last N lines from the provided file,
+// reading from the beginning of the file and keeping track of the last N lines.
+// This is necessary for files that are not seekable (e.g. stdin), but it's less
+// efficient. For an efficient alternative that works on seekable files see
+// readLastLinesBackwards.
+// It returns an error, if any. If no error is encountered, the File object's
+// offset is positioned after the last read location.
+func readLastLinesFromBeginning(inFile *os.File, writer io.Writer, numLines int) error {
+	blkSize := getBlockSize(numLines)
+	// read block by block until EOF and store a reference to the last lines
+	buf := make([]byte, blkSize) // will hold each block of read data
+	var slice []byte             // will hold the final data, after moving line by line
+	foundNewLines := 0
+	for {
+		n, err := io.ReadFull(inFile, buf)
+		if err != nil {
+			if err == io.EOF {
+				break
+			}
+			if err != io.ErrUnexpectedEOF {
+				return err
+			}
+		}
+		// look for newlines and keep a slice starting at the n-th to last line
+		// (no further than numLines)
+		foundNewLines += bytes.Count(buf[:n], []byte{'\n'})
+		slice = append(slice, buf[:n]...) // this is the slice that points to the wanted lines
+		// process the current slice
+		newLinesToSkip := 0
+		if foundNewLines > numLines {
+			newLinesToSkip = foundNewLines - numLines
+		}
+		// if there are more than `numLines` lines, skip them
+		for idx := 0; idx < newLinesToSkip; idx++ {
+			pos := bytes.IndexByte(slice, '\n')
+			// in theory the code below should never panic for out of bounds because
+			// of the checks above, but..
+			if len(slice) < pos+1 {
+				return fmt.Errorf("less bytes than expected were found")
+			}
+			slice = slice[pos+1:]
+			foundNewLines--
+		}
+	}
+	if _, err := writer.Write(slice); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Tail reads the last N lines from the input File and writes them to the Writer.
+// The TailConfig object allows to specify the precise behaviour.
+func Tail(inFile *os.File, writer io.Writer, config TailConfig) error {
+	if config.follow {
+		return fmt.Errorf("follow-mode not implemented yet")
+	}
+	var (
+		err                error
```

Comment:
it's customary in Go not to declare variables until you actually need them. e.g., if you remove the error declaration here and just use

```go
err := readLastLinesBackwards(...
```
that'll do your error decl for you.

Reply:
Absolutely right. This is a leftover of the original code when `readLastLinesBackwards` returned both int64 and error. Fixed in the next commit